src/hvm.rs

```diff
@@ -216,163 +220,151 @@ impl Numb {
-          NEQ => Numb::new_u24((av != bv) as u32),
-          LT  => Numb::new_u24((av <  bv) as u32),
-          GT  => Numb::new_u24((av >  bv) as u32),
-          AND => Numb::new_u24(av & bv),
-          OR  => Numb::new_u24(av | bv),
-          XOR => Numb::new_u24(av ^ bv),
-          _   => unreachable!(),
+          OP_ADD => Numb::new_u24(av.wrapping_add(bv)),
+          OP_SUB => Numb::new_u24(av.wrapping_sub(bv)),
+          FP_SUB => Numb::new_u24(bv.wrapping_sub(av)),
+          OP_MUL => Numb::new_u24(av.wrapping_mul(bv)),
+          OP_DIV => Numb::new_u24(av.wrapping_div(bv)),
+          FP_DIV => Numb::new_u24(bv.wrapping_div(av)),
+          OP_REM => Numb::new_u24(av.wrapping_rem(bv)),
+          FP_REM => Numb::new_u24(bv.wrapping_rem(av)),
+          OP_EQ  => Numb::new_u24((av == bv) as u32),
+          OP_NEQ => Numb::new_u24((av != bv) as u32),
+          OP_LT  => Numb::new_u24((av <  bv) as u32),
+          OP_GT  => Numb::new_u24((av >  bv) as u32),
+          OP_AND => Numb::new_u24(av & bv),
+          OP_OR  => Numb::new_u24(av | bv),
+          OP_XOR => Numb::new_u24(av ^ bv),
+          OP_SHL => Numb::new_u24(av << (bv & 31)),
+          OP_SHR => Numb::new_u24(av >> (bv & 31)),
+          FP_SHL => Numb::new_u24(bv << (av & 31)), 
+          FP_SHR => Numb::new_u24(bv >> (av & 31)),
+          _      => unreachable!(),
         }
       }
-      I24 => {
+      TY_I24 => {
         let av = a.get_i24();
         let bv = b.get_i24();
         match op {
-          ADD => Numb::new_i24(av.wrapping_add(bv)),
-          SUB => Numb::new_i24(av.wrapping_sub(bv)),
-          MUL => Numb::new_i24(av.wrapping_mul(bv)),
-          DIV => Numb::new_i24(av.wrapping_div(bv)),
-          REM => Numb::new_i24(av.wrapping_rem(bv)),
-          EQ  => Numb::new_i24((av == bv) as i32),
-          NEQ => Numb::new_i24((av != bv) as i32),
-          LT  => Numb::new_i24((av <  bv) as i32),
-          GT  => Numb::new_i24((av >  bv) as i32),
-          AND => Numb::new_i24(av & bv),
-          OR  => Numb::new_i24(av | bv),
-          XOR => Numb::new_i24(av ^ bv),
-          _   => unreachable!(),
+          OP_ADD => Numb::new_i24(av.wrapping_add(bv)),
+          OP_SUB => Numb::new_i24(av.wrapping_sub(bv)),
+          FP_SUB => Numb::new_i24(bv.wrapping_sub(av)),
+          OP_MUL => Numb::new_i24(av.wrapping_mul(bv)),
+          OP_DIV => Numb::new_i24(av.wrapping_div(bv)),
+          FP_DIV => Numb::new_i24(bv.wrapping_div(av)),
+          OP_REM => Numb::new_i24(av.wrapping_rem(bv)),
+          FP_REM => Numb::new_i24(bv.wrapping_rem(av)),
+          OP_EQ  => Numb::new_i24((av == bv) as i32),
+          OP_NEQ => Numb::new_i24((av != bv) as i32),
+          OP_LT  => Numb::new_i24((av <  bv) as i32),
+          OP_GT  => Numb::new_i24((av >  bv) as i32),
+          OP_AND => Numb::new_i24(av & bv),
+          OP_OR  => Numb::new_i24(av | bv),
+          OP_XOR => Numb::new_i24(av ^ bv),
+          _      => unreachable!(),
```

Comment:
Do we want to panic here for doing shifts on `i24`? Same question for `f24`.

Reply:
will think about it!